lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1649,169 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
+      auto value = stageResults[i];
+
+      // Get the pipeline register for that result.
+      auto pipelineRegister = pipelineRegisters[i];
+
+      // Get the evaluating group for that value.
+      auto evaluatingGroup = getComponentState().getEvaluatingGroup(value);
+
+      // Remember the final group for this stage result.
+      calyx::GroupOp group;
+
+      // Stitch the register in, depending on whether the group was
+      // combinational or sequential.
+      if (auto combGroup =
+              dyn_cast<calyx::CombGroupOp>(evaluatingGroup.getOperation())) {
+        // Create a sequential group and replace the comb group.
+        PatternRewriter::InsertionGuard g(rewriter);
+        rewriter.setInsertionPoint(combGroup);
+        group = rewriter.create<calyx::GroupOp>(combGroup.getLoc(),
+                                                combGroup.getName());
+        rewriter.cloneRegionBefore(combGroup.getBodyRegion(), group.getBody());
+        group.getBodyRegion().back().erase();
+        rewriter.eraseOp(combGroup);
+
+        // Stitch evaluating group to register.
+        buildAssignmentsForRegisterWrite(getComponentState(), rewriter, group,
+                                         pipelineRegister, value);
+
+        // Register all sources as having been made sequential.
+        for (auto assign : group.getOps<calyx::AssignOp>())
+          getComponentState().addSeqGroupValue(assign.src());
+      } else {
+        // Get the group and register that is temporarily being written to.
+        group = cast<calyx::GroupOp>(evaluatingGroup.getOperation());
+        auto doneOp = group.getDoneOp();
+        auto tempReg =
+            cast<calyx::RegisterOp>(doneOp.src().cast<OpResult>().getOwner());
+        auto tempIn = tempReg.in();
+        auto tempWriteEn = tempReg.write_en();
+
+        // Replace the register write with a write to the pipeline register.
+        for (auto assign : group.getOps<calyx::AssignOp>()) {
+          if (assign.dest() == tempIn)
+            assign.destMutable().assign(pipelineRegister.in());
+          if (assign.dest() == tempWriteEn)
+            assign.destMutable().assign(pipelineRegister.write_en());
+        }
+        doneOp.srcMutable().assign(pipelineRegister.done());
+
+        // Remove the old register completely.
+        rewriter.eraseOp(tempReg);
+      }
+
+      // Replace the stage result uses with the register out.
+      stage.getResult(i).replaceAllUsesWith(pipelineRegister.out());
+
+      // Mark the group for scheduling in the pipeline's block.
+      getComponentState().addBlockScheduleable(stage->getBlock(), group);
+
+      // Add the group to the prologue or epilogue as necessary.
+      if (static_cast<uint64_t>(stage.start()) <= whileOp.II())
+        prologueGroups.push_back(group.sym_nameAttr());
+      if (static_cast<uint64_t>(stage.start()) >= whileOp.II())
+        epilogueGroups.push_back(group.sym_nameAttr());
+    }
```

Comment:
Not super familiar here, but is it intentional here to push to both prologue and epilogue when `stage.start() == whileOp.II()`? If so, can you explain why for my understanding?

Also, DRY nit: save the `static_cast`ed value to a variable, as well as the group's `sym_name`. I had to glance twice to make sure these weren't different.